prober/websocket.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prober
+
+import (
+	"context"
+	"crypto/tls"
+	"log/slog"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/websocket"
+	"github.com/prometheus/blackbox_exporter/config"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+func ProbeWebsocket(ctx context.Context, target string, module config.Module, registry *prometheus.Registry, logger *slog.Logger) (success bool) {
+
+	targetURL, err := url.Parse(target)
+	if err != nil {
+		logger.Error("Could not parse target URL", "err", err)
+		return false
+	}
+
+	logger.Debug("probing websocket", "target", targetURL.String())
+
+	httpStatusCode := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_http_status_code",
+		Help: "Response HTTP status code",
+	})
+	isConnected := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_is_upgraded",
+		Help: "Indicates if the websocket connection was successfully upgraded",
+	})
+
+	registry.MustRegister(isConnected)
+	registry.MustRegister(httpStatusCode)
+
+	dialer := websocket.Dialer{
+		TLSClientConfig: &tls.Config{
```

Comment:
once we use the TLSConfig from `github.com/prometheus/common/config`, please configure more fields so end users can control those.

TLS MinVersion and TLS ServerName seems to be something I can think we need to expose but it also make sense to expose CA cert and client cert configs so folks can configure tls for websocket probes.